vowpalwabbit/options_boost_po.cc

```diff
@@ -144,8 +158,24 @@ void options_boost_po::add_and_parse(const option_group_definition& group)
 
 bool options_boost_po::add_parse_and_check_necessary(const option_group_definition& group)
 {
-  this->add_and_parse(group);
-  return group.check_necessary_enabled(*this) && group.check_one_of();
+  this->internal_add_and_parse(group);
+
```

Comment:
I'm not quite following this bit of logic.

Reply:
There are two ways in which an option group is added:
- `add_and_parse`, or
- `add_parse_and_check_necessary`

The semantic difference here is options added by `add_and_parse` are assumed to be always reachable. Whereas options added with `add_parse_and_check_necessary` are only reachable if the corresponding necessary options of the given group have been supplied.

For example, for CCB `--all_slots_loss` only used if CCB is enabled with `--ccb_explore_adf`.